### PR TITLE
use a table as sentinel value ; check the type of module name argument

### PR DESCRIPTION
**Pull request by tst2005**

### CODE REVIEW

**Comment by pygy** [comment]
Thanks for the patch, but it isn't acceptable as is, because I want to emulate the C code as closely as possible:

```Bash
$ cat >2.lua
print("TWO")
$ lua
Lua 5.2.3  Copyright (C) 1994-2013 Lua.org, PUC-Rio
> require(2)
TWO
>
```

Also:
```Bash
$ cat > loop.lua
require'loop'
$ luajit
LuaJIT 2.0.3 -- Copyright (C) 2005-2014 Mike Pall. http://luajit.org/
JIT: ON CMOV SSE2 SSE3 SSE4.1 fold cse dce fwd dse narrow loop abc sink fuse
> =pcall(require, 'loop')
false	./1.lua:1: loop or previous error loading module '1'
> =package.loaded['loop']
userdata: 0x0100076f60
```

I'd rather not use a mutable value as a fallback. maybe a noop function.

**Comment by tst2005** [comment]
I think the sentinel value should a non-predictive value.
Then `newproxy`, empty table or empty anonymous function should be ok.
But : 
 * newproxy is not documented (planned to be remove ? in Lua 5.3+ ? I don't remember where I read that)
 * since Lua 5.3 empty function is not guarantied to be uniq
 * I usualy use empty table, the only remaining standard value guarantied to be uniq

Do you know if the C code use concat to support require( <number> ).
I think it will be cleaner to check if the type is a number and convert it to string
```lua
name =  type(name) == "number" and tostring(name) or name
```
But if the C part really use concat, your code should be better.

I commit https://github.com/pygy/require.lua/commit/76b176211a2a36a4aacd4b0a007e21ae477020f5 to support numerical modulename.
Now your both sample pass with lua5.1 lua5.2 and luajit.

**Comment by pygy** [comment]
newproxy is only used in v5.1and LuaJIT which is 5.1 compatible. It was removed in Lua 5.2 because in that version, the `__len` metamethod on table is honoured. That was the only use case for proxies.

There's a fallback for the unlikely case that you're using `require.lua` in a v5.1 sandbox that strips `newproxy()`, or if you somehow want to have the 5.1 logic in Lua 5.2 (even less likely for anything beside testing).

Regarding `checkstring`, the current way is indeed a hack. The original `require` uses `luaL_checkstring` and I kinda emulated it using concat, but checking the type (string/num) would be fine too.

File: require.lua [hunk 1]
```diff
@@ -8,25 +8,25 @@ local t_concat = table.concat
...
 
 --- for Lua 5.1
 
 local package, p_loaded = package, package.loaded
 
-local sentinel = newproxy and newproxy() or string.char(1,2,3,4,5)
+local sentinel = {} --newproxy and newproxy() or string.char(1,2,3,4,5)
 
 local function require51 (name)
-    name = checkstring(name)
```
Comment (pygy) [inline require.lua]: You can't get rid of this. `require(2)` looks for the string `"2"` in `package.loaded`.

Comment (tst2005) [inline require.lua]: Oh, You've right.

File: require.lua [hunk 2]
```diff
@@ -8,25 +8,25 @@ local t_concat = table.concat
...
     if p_loaded[name] == sentinel then
         error("loop or previous error loading module '"..name.."'", 2)
     end
+    name = checkstring(name)
```
Comment (pygy) [inline require.lua]: This must happen before `if p_loaded[name] == sentinel`.

**Comment by pygy** [comment]
Rather than micromanaging you to get what I want as a fix, I wrote it myself... Is that patch ok for you?

File: require.lua (hunk 2 above)
Comment (tst2005) [inline require.lua]: Exact!

**Comment by tst2005** [comment]
Your patch seems good.
I see only one minor improvment, you should use error("error message", <b>2</b>)  to stop the stacktrace at the try to index or update the sentinel.
```lua
-local function errhandler() error("the require sentinel can't be indexed or updated") end
+local function errhandler() error("the require sentinel can't be indexed or updated", 2) end
```

File: require.lua (hunk 2 above)
Comment (tst2005) [inline require.lua]: Exact !

**Comment by tst2005** [comment]
Your patch seems good.
I see only one minor improvment, you should use error("error message", <b>2</b>)  to stop the stacktrace at the try to index or update the sentinel.

```lua
-local function errhandler() error("the require sentinel can't be indexed or updated") end
+local function errhandler() error("the require sentinel can't be indexed or updated", 2) end
```

**Comment by pygy** [comment]
This is indeed better :-)

**Comment by pygy** [comment]
#4 is merged. Now I'll have to remember how to do the Luarocks dance.

**Pull request closed without merging**